### DISCUSSION TITLE
How critical is 'fetch' in terms of caching, revalidating, etc.

### DISCUSSION THREAD

**Discussion by Callumk7** [vercel/next.js#52772] [App Router]
Hey, 

Most of the documentation suggests that the optimisations provided by the framework specifically rely on data being fetched in server components with fetch. I am wondering how other data collection methods fit in with Next, for example using an ORM like Prisma. Will a server component that gets data using prisma follow the same rules? Will I be able to set 'force-dynamic' and 'revalidate - 0' if there are no fetch functions? Or is the correct pattern here to create an api route which handles and returns the data with prisma, and the server component should fetch from there?

**Comment by icyJoseph** [comment]
- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#data-fetching-without-fetch

> Any data fetching libraries that do not use fetch directly will not affect caching of a route, and will be static or dynamic depending on the route segment.

It really depends... I've for example used Meilisearch as DB, and their JS SDK uses network calls to communicate with the DB so that gets automatically patched. 

>  Or is the correct pattern here to create an api route which handles and returns the data with prisma, and the server component should fetch from there?

Not really, that also has the implication that those are publicly reachable. 

Also, they won't be available during build time, so you run into a catch-22 kind of situation, no server to serve the API request, because we are in the middle of doing a build.

I guess there'll be some solution coming out from Nextjs or the various SDKs.

There's a hacky way to invoke fetch, make it fail and collect data through an SDK, but I'd not recommend that. I would prefer a first party kind of solution.

**Reply by Callumk7** [reply]
Thanks for the reply. I am pretty new to SSR style apps, and I feel like I am falling at the first hurdle here. 

I have a pretty normal behaviour that I would like to achieve:
1. user is on the collection page (server component), which collects collection data via prisma. 
2. user navigates to the search page, finds another item to add to the collection, and does so. 
3. the user navigates back to the collection page, which is 're-run' such that prisma collects the updated data. 

So far, I cannot get 3. to NOT display a cached version of the state of the page at 1. I have tried using
export const revalidate = 0;
export const dynamic = "force-dynamic";
on the collection page, but this seems to have no effect. 

Really not sure what the correct pattern would be here, as this must be a super common case.

**Reply by icyJoseph** [reply]
Yeah the client side Server Component cache is a bit unintuitive right now.

What you can do in step 2, is use a "Server Action", which revalidates the path where you see the collection. 

I had such issue with a cookie: https://github.com/vercel/next.js/discussions/51488#discussioncomment-6216934

And a somewhat related answer too: https://github.com/vercel/next.js/discussions/52760#discussioncomment-6464426

**Reply by Callumk7** [reply]
Thanks @icyJoseph 
I had read about the `revalidatePath` function, and its recommendation to use server actions, but to be honest I had hoped to avoid it due to the fact that its an alpha feature. 

I have however found a semi-solution: I set `export const revalidate = 1;` Which (seems) to actually have the desired effect of revalidating the route (at least every second, if not on every request -- but thats *probably* fine?)

Really strange how the revalidate = 0; as recommended from the docs doesn't seem to work as advertised.

**Reply by jviall** [reply]
> >Or is the correct pattern here to create an api route which handles and returns the data with prisma, and the server component should fetch from there?
>
> Not really, that also has the implication that those are publicly reachable.

@icyJoseph do you mean it shouldn't be the correct pattern because exposing it to the public (as opposed to being a server side function that is called directly from a server component) is an unnecessary exposure of vulnerability?

My impression is that Route Handlers are the preferred way to interact with the server from a component, otherwise I have a hard time understanding when to use a Route Handler and when to just have components directly call a server-only function that fetches data--what is the impact of one v.s. the other? documentation is severely lacking here, and that's not even touching on all the complexities of the various caching options.

In my case, I'm implementing server-side code in a Controller/Service/Repo pattern where my Route Handlers act as "Controllers"--entry-points into my Next App's API. Repos are responsible for 1 database each; Services implement and orchestrate functionality using repos and other services; Route Handlers simply take a request, pass it to the Service method that implements the relevant functionality, and returns the response.

but I'm running into all sorts of caching problems.
